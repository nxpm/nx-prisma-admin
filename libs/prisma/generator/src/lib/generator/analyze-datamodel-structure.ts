import { DMMF } from '@prisma/generator-helper'
import { DatamodelFieldMap } from './datamodel-field-map'
import { DatamodelStructure } from './datamodel-structure'
import { sortBy } from './helpers'

export function getModelFieldMap(model: DMMF.Model): DatamodelFieldMap {
  const fields = model.fields
  const enums = fields.filter((f) => f.kind === 'enum')
  const scalars = fields.filter((f) => f.kind === 'scalar')
  const relations = fields.filter((f) => f.kind === 'object')

  // Don't expose the scalar fields that store the relation
  const relationFieldIs = []
  relations.forEach((item) => {
    if (item.relationFromFields?.length) {
      relationFieldIs.push(...item.relationFromFields)
    }
  })

  return {
    model,
    enums,
    scalars: scalars.filter((scalar) => !relationFieldIs.includes(scalar.name)),
    relations,
  }
}

export function analyzeDatamodelStructure(dmmf: DMMF.Document): DatamodelStructure {
  console.log(JSON.stringify(dmmf.schema.inputObjectTypes.prisma, null, 2))
  const enums = sortBy(dmmf.datamodel.enums, 'name')
  const models = sortBy(dmmf.datamodel.models, 'name')
  const enumMap = enums.reduce((acc, curr) => ({ ...acc, [curr.name]: curr }), {})
  const modelMap = models.reduce((acc, curr) => ({ ...acc, [curr.name]: curr }), {})
  const modelFieldMap = models.reduce((acc, curr) => ({ ...acc, [curr.name]: getModelFieldMap(curr) }), {})

  return {
    enumMap,
    modelMap,
    modelFieldMap,
  }
}
