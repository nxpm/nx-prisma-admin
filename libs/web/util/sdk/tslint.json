{
  "extends": "../../../../tslint.json",
  "rules": {
    "directive-selector": [true, "attribute", "sdk", "camelCase"],
    "component-selector": [true, "element", "sdk", "kebab-case"]
  },
  "linterOptions": {
    "exclude": ["!**/*"]
  }
}
