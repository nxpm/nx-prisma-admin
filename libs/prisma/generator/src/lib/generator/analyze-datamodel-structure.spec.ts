import { DMMF } from '@prisma/generator-helper'
import { getDMMF } from '@prisma/sdk'
import { readFileSync } from 'fs'
import { analyzeDatamodelStructure } from './analyze-datamodel-structure'

const datamodel = readFileSync(`${__dirname}/test-schema.prisma`).toString()

fdescribe('analyze-datamodel-structure', () => {
  let dmmf: DMMF.Document

  beforeEach(async () => {
    dmmf = await getDMMF({ datamodel })
  })

  it('should create basic model definition', () => {
    const schema = analyzeDatamodelStructure(dmmf)

    expect(schema.enumMap).toMatchSnapshot()
    expect(schema.modelMap).toMatchSnapshot()
    expect(schema.modelFieldMap).toMatchSnapshot()
  })
})
