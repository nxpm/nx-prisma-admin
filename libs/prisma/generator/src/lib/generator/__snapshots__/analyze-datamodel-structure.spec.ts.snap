// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyze-datamodel-structure should create basic model definition 1`] = `
Object {
  "Role": Object {
    "dbName": null,
    "name": "Role",
    "values": Array [
      Object {
        "dbName": null,
        "name": "USER",
      },
      Object {
        "dbName": null,
        "name": "ADMIN",
      },
    ],
  },
}
`;

exports[`analyze-datamodel-structure should create basic model definition 2`] = `
Object {
  "Post": Object {
    "dbName": null,
    "fields": Array [
      Object {
        "default": Object {
          "args": Array [],
          "name": "autoincrement",
        },
        "hasDefaultValue": true,
        "isGenerated": false,
        "isId": true,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "id",
        "type": "Int",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "title",
        "type": "String",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "object",
        "name": "user",
        "relationFromFields": Array [
          "userId",
        ],
        "relationName": "PostToUser",
        "relationOnDelete": "NONE",
        "relationToFields": Array [
          "id",
        ],
        "type": "User",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": true,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "userId",
        "type": "Int",
      },
    ],
    "idFields": Array [],
    "isEmbedded": false,
    "isGenerated": false,
    "name": "Post",
    "uniqueFields": Array [],
    "uniqueIndexes": Array [],
  },
  "User": Object {
    "dbName": null,
    "fields": Array [
      Object {
        "default": Object {
          "args": Array [],
          "name": "autoincrement",
        },
        "hasDefaultValue": true,
        "isGenerated": false,
        "isId": true,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "id",
        "type": "Int",
      },
      Object {
        "default": Object {
          "args": Array [],
          "name": "now",
        },
        "hasDefaultValue": true,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "createdAt",
        "type": "DateTime",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": true,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "email",
        "type": "String",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "weight",
        "type": "Float",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "is18",
        "type": "Boolean",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "name",
        "type": "String",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": true,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "successorId",
        "type": "Int",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "object",
        "name": "successor",
        "relationFromFields": Array [
          "successorId",
        ],
        "relationName": "BlogOwnerHistory",
        "relationOnDelete": "NONE",
        "relationToFields": Array [
          "id",
        ],
        "type": "User",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "object",
        "name": "predecessor",
        "relationFromFields": Array [],
        "relationName": "BlogOwnerHistory",
        "relationOnDelete": "NONE",
        "relationToFields": Array [],
        "type": "User",
      },
      Object {
        "default": "USER",
        "hasDefaultValue": true,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "enum",
        "name": "role",
        "type": "Role",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": true,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "object",
        "name": "posts",
        "relationFromFields": Array [],
        "relationName": "PostToUser",
        "relationOnDelete": "NONE",
        "relationToFields": Array [],
        "type": "Post",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": true,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "keywords",
        "type": "String",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "biography",
        "type": "Json",
      },
    ],
    "idFields": Array [],
    "isEmbedded": false,
    "isGenerated": false,
    "name": "User",
    "uniqueFields": Array [],
    "uniqueIndexes": Array [],
  },
}
`;

exports[`analyze-datamodel-structure should create basic model definition 3`] = `
Object {
  "Post": Object {
    "enums": Array [],
    "model": Object {
      "dbName": null,
      "fields": Array [
        Object {
          "default": Object {
            "args": Array [],
            "name": "autoincrement",
          },
          "hasDefaultValue": true,
          "isGenerated": false,
          "isId": true,
          "isList": false,
          "isReadOnly": false,
          "isRequired": true,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "id",
          "type": "Int",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": true,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "title",
          "type": "String",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": false,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "object",
          "name": "user",
          "relationFromFields": Array [
            "userId",
          ],
          "relationName": "PostToUser",
          "relationOnDelete": "NONE",
          "relationToFields": Array [
            "id",
          ],
          "type": "User",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": true,
          "isRequired": false,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "userId",
          "type": "Int",
        },
      ],
      "idFields": Array [],
      "isEmbedded": false,
      "isGenerated": false,
      "name": "Post",
      "uniqueFields": Array [],
      "uniqueIndexes": Array [],
    },
    "relations": Array [
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "object",
        "name": "user",
        "relationFromFields": Array [
          "userId",
        ],
        "relationName": "PostToUser",
        "relationOnDelete": "NONE",
        "relationToFields": Array [
          "id",
        ],
        "type": "User",
      },
    ],
    "scalars": Array [
      Object {
        "default": Object {
          "args": Array [],
          "name": "autoincrement",
        },
        "hasDefaultValue": true,
        "isGenerated": false,
        "isId": true,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "id",
        "type": "Int",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "title",
        "type": "String",
      },
    ],
  },
  "User": Object {
    "enums": Array [
      Object {
        "default": "USER",
        "hasDefaultValue": true,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "enum",
        "name": "role",
        "type": "Role",
      },
    ],
    "model": Object {
      "dbName": null,
      "fields": Array [
        Object {
          "default": Object {
            "args": Array [],
            "name": "autoincrement",
          },
          "hasDefaultValue": true,
          "isGenerated": false,
          "isId": true,
          "isList": false,
          "isReadOnly": false,
          "isRequired": true,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "id",
          "type": "Int",
        },
        Object {
          "default": Object {
            "args": Array [],
            "name": "now",
          },
          "hasDefaultValue": true,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": true,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "createdAt",
          "type": "DateTime",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": true,
          "isUnique": true,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "email",
          "type": "String",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": false,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "weight",
          "type": "Float",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": false,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "is18",
          "type": "Boolean",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": false,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "name",
          "type": "String",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": true,
          "isRequired": false,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "successorId",
          "type": "Int",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": false,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "object",
          "name": "successor",
          "relationFromFields": Array [
            "successorId",
          ],
          "relationName": "BlogOwnerHistory",
          "relationOnDelete": "NONE",
          "relationToFields": Array [
            "id",
          ],
          "type": "User",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": true,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "object",
          "name": "predecessor",
          "relationFromFields": Array [],
          "relationName": "BlogOwnerHistory",
          "relationOnDelete": "NONE",
          "relationToFields": Array [],
          "type": "User",
        },
        Object {
          "default": "USER",
          "hasDefaultValue": true,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": true,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "enum",
          "name": "role",
          "type": "Role",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": true,
          "isReadOnly": false,
          "isRequired": false,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "object",
          "name": "posts",
          "relationFromFields": Array [],
          "relationName": "PostToUser",
          "relationOnDelete": "NONE",
          "relationToFields": Array [],
          "type": "Post",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": true,
          "isReadOnly": false,
          "isRequired": false,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "keywords",
          "type": "String",
        },
        Object {
          "hasDefaultValue": false,
          "isGenerated": false,
          "isId": false,
          "isList": false,
          "isReadOnly": false,
          "isRequired": true,
          "isUnique": false,
          "isUpdatedAt": false,
          "kind": "scalar",
          "name": "biography",
          "type": "Json",
        },
      ],
      "idFields": Array [],
      "isEmbedded": false,
      "isGenerated": false,
      "name": "User",
      "uniqueFields": Array [],
      "uniqueIndexes": Array [],
    },
    "relations": Array [
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "object",
        "name": "successor",
        "relationFromFields": Array [
          "successorId",
        ],
        "relationName": "BlogOwnerHistory",
        "relationOnDelete": "NONE",
        "relationToFields": Array [
          "id",
        ],
        "type": "User",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "object",
        "name": "predecessor",
        "relationFromFields": Array [],
        "relationName": "BlogOwnerHistory",
        "relationOnDelete": "NONE",
        "relationToFields": Array [],
        "type": "User",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": true,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "object",
        "name": "posts",
        "relationFromFields": Array [],
        "relationName": "PostToUser",
        "relationOnDelete": "NONE",
        "relationToFields": Array [],
        "type": "Post",
      },
    ],
    "scalars": Array [
      Object {
        "default": Object {
          "args": Array [],
          "name": "autoincrement",
        },
        "hasDefaultValue": true,
        "isGenerated": false,
        "isId": true,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "id",
        "type": "Int",
      },
      Object {
        "default": Object {
          "args": Array [],
          "name": "now",
        },
        "hasDefaultValue": true,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "createdAt",
        "type": "DateTime",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": true,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "email",
        "type": "String",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "weight",
        "type": "Float",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "is18",
        "type": "Boolean",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "name",
        "type": "String",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": true,
        "isReadOnly": false,
        "isRequired": false,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "keywords",
        "type": "String",
      },
      Object {
        "hasDefaultValue": false,
        "isGenerated": false,
        "isId": false,
        "isList": false,
        "isReadOnly": false,
        "isRequired": true,
        "isUnique": false,
        "isUpdatedAt": false,
        "kind": "scalar",
        "name": "biography",
        "type": "Json",
      },
    ],
  },
}
`;
